File: python/algorithms/binary_search_tree.py
"""Binary search tree example
Copyright 2016, Sjors van Gelderen
"""

"""Binary search tree
Copyright 2016, Sjors van Gelderen
"""
class BST:
    def __init__(self, _parent, _key):
        self.parent = _parent
        self.left = None
        self.right = None
        self.key = _key

    # Complexity: O(h) where h = BST height
    def insert(self, _value):        
        try:
            if self.parent == None:
                print("Starting insertion of {}".format(_value))
            
            if _value < self.key:
                if self.left == None:
                    print("Inserting node with value {} left of {}".format(_value, self.key))
                    self.left = BST(self, _value)
                else:
                    print("Moving left from {}".format(self.key))
                    self.left.insert(_value)
            elif _value > self.key:
                if self.right == None:
                    print("Inserting node with value {} right of {}".format(_value, self.key))
                    self.right = BST(self, _value)
                else:
                    print("Moving right from {}".format(self.key))
                    self.right.insert(_value)
            else:
                raise Exception("Duplicate entry, ignoring insert!")
        except:
            pass

    # Complexity: O(h) where h = BST height
    def find(self, _value):
        if self.parent == None:
            print("Looking for {}".format(_value))
        
        if _value < self.key:
            if self.left == None:
                print("Could not find value!")
            else:
                print("Moving left from {}".format(self.key))
                self.left.find(_value)
        elif _value > self.key:
            if self.right == None:
                print("Could not find value!")
            else:
                print("Moving right from {}".format(self.key))
                self.right.find(_value)
        else:
            print("Found value!")
                
# Main program logic
def program():
    tree = BST(None, 49)
    tree.insert(52)
    tree.insert(3)
    tree.insert(999)
    tree.insert(33)
    
    tree.find(23)
    tree.find(55)
    tree.find(3)
    tree.find(999)

program()

File: c/algorithms/binary_search_tree.c
/*
  Binary search tree example
  Copyright 2016, Sjors van Gelderen
*/

#include <stdlib.h>
#include <stdio.h>

// Binary search tree data structure
typedef struct BST
{
    struct BST* parent;
    struct BST* left;
    struct BST* right;
    int key;
} BST;


// Function prototypes
int print_array(int _n, int _array[_n]);
BST* bst_create();
void bst_insert(BST* _tree, int _key);
void bst_find(BST* _tree, int _key);

// Prints an array
int print_array(int _n, int _array[_n])
{
    int i;

    if(_n <= 0)
    {
	printf("Array is of invalid length!\n");
	return EXIT_FAILURE;
    }
    
    printf("[");
    
    for(i = 0; i < _n; i++)
    {
	if(i < _n -1)
	{
	    printf("%d, ", _array[i]);
	}
	else
	{
	    printf("%d", _array[i]);
	}
    }

    printf("]\n");
    return EXIT_SUCCESS;
}

// Creates a BST and returns a pointer to it
BST* bst_create(BST* _parent, int _key)
{
    BST* tree = malloc(sizeof(BST));
    if(tree == NULL)
    {
	printf("Error, malloc for BST failed!\n");
	return NULL;
    }
  
    tree->parent = _parent;
    tree->key = _key;
    return tree;
}

/*
  Inserts a node into the BST
  Complexity: O(h) where h = height of BST
*/
void bst_insert(BST* _tree, int _key)
{
    if(_tree->parent == NULL)
    {
	printf("Starting insertion of %d\n", _key);
    }
    
    if(_key < _tree->key)
    {
	if(_tree->left == NULL)
	{
	    printf("Inserting node with key %d left of %d\n", _key, _tree->key);
	    _tree->left = bst_create(_tree, _key);
	}
	else
	{
	    printf("Moving left from %d\n", _tree->key);
	    bst_insert(_tree->left, _key);
	}
    }
    else if(_key > _tree->key)
    {
	if(_tree->right == NULL)
	{
	    printf("Inserting node with key %d right of %d\n", _key, _tree->key);
	    _tree->right = bst_create(_tree, _key);
	}
	else
	{
	    printf("Moving right from %d\n", _tree->key);
	    bst_insert(_tree->right, _key);
	}
    }
    else
    {
	printf("Duplicate entry, ignoring insert!\n");
	return;
    }
}

/*
  Find a node in the BST
  Complexity: O(h) where h = height of BST
*/
void bst_find(BST* _tree, int _key)
{
    if(_tree->parent == NULL)
    {
	printf("Looking for %d\n", _key);
    }

    if(_key < _tree->key)
    {
	if(_tree->left == NULL)
	{
	    printf("Could not find key!\n");
	}
	else
	{
	    printf("Moving left from %d\n", _tree->key);
	    bst_find(_tree->left, _key);
	}
    }
    else if(_key > _tree->key)
    {
	if(_tree->right == NULL)
	{
	    printf("Could not find key!\n");
	}
	else
	{
	    printf("Moving right from %d\n", _tree->key);
	    bst_find(_tree->right, _key);
	}
    }
    else
    {
	printf("Found key!\n");
    }
}

// Main program logic
int main()
{
    int array[5] = {1, 3, 2, 7, 6};

    BST tree;
    bst_insert(&tree, 49);
    bst_insert(&tree, 52);
    bst_insert(&tree, 3);
    bst_insert(&tree, 999);
    bst_insert(&tree, 33);

    bst_find(&tree, 23);
    bst_find(&tree, 55);
    bst_find(&tree, 3);
    bst_find(&tree, 999);
    
    return EXIT_SUCCESS;
}
